File: backend/app/api/v1/companies.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.dependencies import PermissionChecker
from app.models.user import User
from app.schemas.company import CompanyCreate, CompanyRead, CompanyUpdate
from app.services.company import (
    create_company,
    get_company,
    list_companies,
    update_company,
)

router = APIRouter(prefix="/companies", tags=["Companies"])


@router.get("", response_model=dict)
async def list_companies_endpoint(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(PermissionChecker("admin.view")),
):
    result = await list_companies(db, page=page, page_size=page_size)
    result["items"] = [CompanyRead.model_validate(c) for c in result["items"]]
    return result


@router.post("", response_model=CompanyRead, status_code=201)
async def create_company_endpoint(
    body: CompanyCreate,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(PermissionChecker("admin.create")),
):
    company = await create_company(db, body)
    return CompanyRead.model_validate(company)


@router.get("/{company_id}", response_model=CompanyRead)
async def get_company_endpoint(
    company_id: int,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(PermissionChecker("admin.view")),
):
    company = await get_company(db, company_id)
    return CompanyRead.model_validate(company)


@router.patch("/{company_id}", response_model=CompanyRead)
async def update_company_endpoint(
    company_id: int,
    body: CompanyUpdate,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(PermissionChecker("admin.edit")),
):
    company = await update_company(db, company_id, body)
    return CompanyRead.model_validate(company)

File: frontend/src/services/companies.ts
import api from "./api";
import type { Company } from "@/types/company";
import type { PaginatedResponse } from "@/types/api";

export interface CompanyListParams {
  [key: string]: unknown;
  page?: number;
  page_size?: number;
  search?: string;
  is_active?: boolean;
}

export async function listCompanies(params: CompanyListParams = {}): Promise<PaginatedResponse<Company>> {
  const { data } = await api.get("/companies", { params });
  return data;
}

export async function getCompany(id: number): Promise<Company> {
  const { data } = await api.get(`/companies/${id}`);
  return data;
}

export async function createCompany(body: Partial<Company>): Promise<Company> {
  const { data } = await api.post("/companies", body);
  return data;
}

export async function updateCompany(id: number, body: Partial<Company>): Promise<Company> {
  const { data } = await api.patch(`/companies/${id}`, body);
  return data;
}

export async function toggleCompanyStatus(id: number): Promise<Company> {
  const { data } = await api.post(`/companies/${id}/toggle-status`);
  return data;
}
